[export] Various fixes to .module()

angelayi opened the pull request:
Summary: While turning on .module() for all the export tests, I uncovered some bugs with .module() and while fixing them I ended up rewriting some of the code... Some of the bugs were: 

* bad kwargs support on the unlifted module
* no support for user input mutations
* (at the commit hash i was working off of) no support for custom objects
* there were no tests on unlifting weights from cond/map submodules

Test Plan: CI

Differential Revision: D53075380




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

suo commented on torch/_export/utils.py:
```diff
@@ -13,20 +13,33 @@
 def _check_input_constraints_pre_hook(self, *args, **kwargs):
-    flat_args, _ = tree_flatten(args)
+    try:
+        args = fx_pytree.tree_flatten_spec(
```
avoid tree_flatten_spec. It's a weird forked thing and doesn't support a bunch of features, notably including custom pytrees.

suo commented on torch/export/_unlift.py:
```diff
@@ -1,273 +1,301 @@
-                param_buffer_name_to_corrected_name[name] = name.replace(".", "_")
-            else:
-                new_gm.register_buffer(name, value)
+            new_gm.register_buffer(name.replace(".", "_"), value)
```
nice, this code is so much cleaner!

suo commented on torch/export/_unlift.py:
```diff
@@ -1,273 +1,301 @@
+            new_gm.register_parameter(name.replace(".", "_"), value)
 
-    return inp_pos_to_param_buffer_name
+    assert hasattr(graph_signature, "lifted_tensor_constants")
```
this check was for the compile_fx path, I don't think it's necessary anymore

suo commented on torch/export/_unlift.py:
```diff
@@ -1,273 +1,301 @@
+    Find the all the buffers and inputs that were mutated and insert copy_
+    operators to reflect mutations.
+    """
+    output_node = list(gm.graph.nodes)[-1]
```
I don't think this is guaranteed; while graphs generated by make_fx/dynamo almost certainly have this property, I remember someone tried to assert this in fbcode and it was tripped. We should probably just search for output to be safe.

suo commented on torch/export/_unlift.py:
```diff
@@ -1,273 +1,301 @@
+    """
+    output_node = list(gm.graph.nodes)[-1]
+    assert output_node.op == "output"
+    outputs = pytree.tree_flatten(output_node.args)[0]
```
I don't quite understand this maneuver lol. If the graph returned a tuple of outputs, wouldn't this flatten it and take only the first element? Maybe I'm misunderstanding the semantics of the output node.

angelayi replied:
The reasoning is that for dynamo returned graph modules, the output is always a singleton tuple, with the first element containing the actual outputs. But for cond submodules the output doesn't follow this scheme -- it could be a single node, or a tuple of nodes.... I remember asking @ydwu4 about this and at the time it was a little hard to change it to follow the tuple output scheme.

suo replied:
I guess my question is: if the node looks like `output_node.args=((1, 2, 3, 4), )`, wouldn't `tree_flatten(output_node.args)[0]` return `1`?

angelayi replied:
Oh, no. tree_flatten returns the flat list and the tree spec (ex. `([1, 2, 3, 4], TreeSpec(Tuple, ...))`). So indexing [0] gives you the flat list.

suo replied:
OH lol yeah that makes sense

suo commented on torch/export/_unlift.py:
```diff
@@ -1,273 +1,301 @@
-    param_buffer_name_to_corrected_name = {}
-    constant_name_to_corrected_name = {}
-
+def _register_attrs_to_new_gm(new_gm, graph_signature, state_dict, constants):
```
maybe add types to this one too?

suo reviewed (changes requested):
This is awesome :) Love the docs/static types added!

One request: could you expand the description a little bit to mention what the bugs were that you fixed? It would help me understand the intent of the new behaviors better.

I added some small comments inline as well.

suo commented on torch/export/_unlift.py:
```diff
@@ -1,273 +1,322 @@
+    state_dict: Dict[str, Any],
+    constants: Dict[str, Any],
+) -> None:
     for name, value in state_dict.items():
```
one subtle thing that I noticed: for these, we are using the state dict as the source of truth for what is available, iterating over it and looking things up in the graph signature.

More properly it should be the other way around: we should treat the graph signature as the source of truth, and look up things in the state dict.

Why does this matter? Because the state dict comes from the user, so it could be wrong or have the wrong keys. The graph signature is our actual ground truth about what should go in the graph and not. 

So we should trust only the graph signature, and also double-check that the provided state dict matches. That includes catching the case where there is an extraneous key in the state dict, which in this code is passed through silently. 

This checking doesn't have to be here (doesn't have to be in this PR at all), maybe it should be earlier in the process. But something I noticed

suo replied:
actually, I think I'll address this in a followup to tighten a few screws around this pass (including the buffer renaming stuff mentioned ovr chat). So you can ignore this feedback for this PR

suo commented on test/export/test_export.py:
```diff
@@ -1680,6 +1674,67 @@ def forward(self, x):
+                1
+            )
+
+    def test_map_buffers(self):
```
curious; was there a motivation for including this specific code pattern?

One thing I realized while basing some work off your diff is that I think we should ban this kind of code. In particular, code that references modules/parameters/buffers that are not in the module hierarchy.

If you define this module in eager, `param` and `buffer` will not be in `M`'s state dict, and `m1` will not be in the `children()`. All of this is very weird and unlikely to lead to correct results if you try to use this module for anything. So my claim is we should just error here.

angelayi replied:
Oops sorry I missed this. No specific motivation for this code pattern. I just wanted a test case that uses a param/buffer inside map 😅

angelayi committed:
[export] Various fixes to .module() (#118272)

Summary:

While turning on .module() for all the export tests, I uncovered some bugs with .module() and while fixing them I ended up rewriting some of the code... Some of the bugs were: 

* bad kwargs support on the unlifted module
* no support for user input mutations
* (at the commit hash i was working off of) no support for custom objects
* there were no tests on unlifting weights from cond/map submodules

Test Plan: CI

Reviewed By: suo

Differential Revision: D53075380
